Sources/AWSLambdaRuntimeCore/Documentation.docc/Proposals/0001-v2-api.md

```diff
@@ -0,0 +1,667 @@
+let postgresClient = PostgresClient()
+
+/// Instantiate LambdaRuntime with a closure handler implementing the business logic of the Lambda function
+let runtime = LambdaRuntime { (event: Input, context: LambdaContext) in
+    /// Use initialized service within the handler
+    try await postgresClient.query(...)
+}
+
+/// Use ServiceLifecycle to manage the initialization and termination
+/// of the services as well as the LambdaRuntime
+let serviceGroup = ServiceGroup(
+    services: [postgresClient, runtime],
+    configuration: .init(gracefulShutdownSignals: [.sigterm]),
+    logger: logger
+)
+try await serviceGroup.run()
+```
+
+### Simplifying Codable support
+
+A detailed explanation is provided in the **Codable Support** section. In short, much of the boilerplate code defined for
+each handler protocol in `Lambda+Codable` and `Lambda+String` will be replaced with a single `LambdaCodableAdapter`
+class.
+
+This adapter class is generic over (1) any handler conforming to `LambdaHandler`, (2) the user-specified input and
+output types, and (3) any decoder and encoder conforming to `LambdaEventDecoder` and `LambdaOutputDecoder`. The adapter
+will wrap the underlying handler with encoding/decoding logic.
+
+### Simplifying the handler protocols
+
+There are four different handler protocols in the current API (`SimpleLambdaHandler`, `LambdaHandler`,
+`EventLoopLambdaHandler`, `ByteBufferLambdaHandler`). As noted in the **Current Limitations** section, the ease-of-use
+varies for each handler protocol and users may not be able to easily determine which protocol best serves their use-case
+without spending time digging into the library. To reduce this problem and provide users with clear-cut options, we
+propose replacing all of the existing handler protocols with just two: `LambdaHandler` and `StreamingLambdaHandler`.
+Both will be explained in the **Detailed Solution** section.
+
+## Detailed Solution
+
+Below are explanations for all types that we want to use in AWS Lambda Runtime v2.
+
+### LambdaResponseWriter
+
+We will introduce a new `LambdaResponseWriter` protocol. It is used in the new `StreamingLambdaHandler` (defined below),
+which is the new base protocol for the `LambdaRuntime` (defined below as well).
+
+```swift
+/// A writer object to write the Lambda response stream into
+public protocol LambdaResponseWriter: ~Copyable {
+    /// Write a response part into the stream
+    func write(_ buffer: ByteBuffer) async throws
+    /// End the response stream
+    consuming func finish()
+    /// Write a response part into the stream and end the response stream
+    consuming func writeAndFinish(_ buffer: ByteBuffer) async throws
+    /// Report an error in the response stream
+    consuming func reportError(_ error: any Error) async throws
+}
+```
+
+### LambdaContext
+
+`LambdaContext` will be largely unchanged, but the `eventLoop` property as well as the `allocator` property (of type
+`ByteBufferAllocator`) will be removed.
```

Comment:
I think it's fine to remove this. The original idea was that you could use more efficient allocators for ByteBuffers but we never actually used that. Swift itself is still not able to use alternative allocators (except if you globally set `malloc`/`free` to a different one) so I don't think this was as useful as we thought it would be (outside of ByteBuffer tests).